ts/client/src/client.ts

```diff
@@ -3374,64 +3372,230 @@ export class MangoClient {
+    );
+  }
+
+  public async tcsStopLossOnBorrow(
+    group: Group,
+    account: MangoAccount,
+    sellBank: Bank,
+    buyBank: Bank,
+    thresholdPriceUi: number,
+    thresholdPriceInSellPerBuyToken: boolean,
+    maxBuyUi: number | null,
+    maxSellUi: number | null,
+    pricePremium: number | null,
+    allowCreatingDeposits: boolean | null,
+    allowCreatingBorrows: boolean | null,
+    expiryTimestamp: number | null,
+  ): Promise<TransactionSignature> {
+    if (maxBuyUi == null && account.getTokenBalanceUi(sellBank) > 0) {
+      throw new Error(
+        `Only allowed to set stop loss on borrows! Current balance ${account.getTokenBalanceUi(
+          buyBank,
+        )}`,
+      );
+    }
+
+    return await this.tokenConditionalSwapCreate(
+      group,
+      account,
+      sellBank,
+      buyBank,
+      thresholdPriceUi,
+      thresholdPriceInSellPerBuyToken,
+      maxBuyUi ?? -account.getTokenBalanceUi(buyBank),
+      maxSellUi ?? Number.MAX_SAFE_INTEGER,
+      'StopLossOnBorrow',
+      pricePremium,
+      allowCreatingDeposits ?? false,
+      allowCreatingBorrows ?? false,
+      expiryTimestamp,
+    );
+  }
+
+  public async tokenConditionalSwapCreate(
+    group: Group,
+    account: MangoAccount,
+    sellBank: Bank,
+    buyBank: Bank,
+    thresholdPriceUi: number,
+    thresholdPriceInSellPerBuyToken: boolean,
+    maxBuyUi: number,
+    maxSellUi: number,
+    tcsIntention:
+      | 'TakeProfitOnDeposit'
+      | 'StopLossOnDeposit'
+      | 'TakeProfitOnBorrow'
+      | 'StopLossOnBorrow'
+      | null,
+    pricePremium: number | null,
+    allowCreatingDeposits: boolean,
+    allowCreatingBorrows: boolean,
+    expiryTimestamp: number | null,
+  ): Promise<TransactionSignature> {
+    const maxBuy = toNative(maxBuyUi, buyBank.mintDecimals);
+    const maxSell = toNative(maxSellUi, sellBank.mintDecimals);
+
+    if (!thresholdPriceInSellPerBuyToken) {
+      thresholdPriceUi = 1 / thresholdPriceUi;
+    }
+
+    let lowerLimit, upperLimit;
+    const thresholdPrice = toNativeSellPerBuyTokenPrice(
+      thresholdPriceUi,
+      sellBank,
+      buyBank,
     );
-    pricePremiumFraction =
-      pricePremiumFraction > 0 ? pricePremiumFraction : 0.3;
+    const sellTokenPerBuyTokenPrice = buyBank.price
+      .div(sellBank.price)
+      .toNumber();
+
+    if (
+      tcsIntention == 'TakeProfitOnDeposit' ||
+      tcsIntention == 'StopLossOnBorrow' ||
+      (tcsIntention == null && thresholdPrice > sellTokenPerBuyTokenPrice)
+    ) {
+      lowerLimit = thresholdPrice;
+      upperLimit = Number.MAX_SAFE_INTEGER;
+    } else {
+      lowerLimit = 0;
+      upperLimit = thresholdPrice;
+    }
+
     const expiryTimestampBn =
       expiryTimestamp !== null ? new BN(expiryTimestamp) : U64_MAX_BN;
 
+    pricePremium = group.getPriceImpactByTokenIndex(
+      sellBank.tokenIndex,
+      maxSellUi * sellBank.uiPrice,
+    );
+    const pricePremiumFraction = pricePremium > 0 ? pricePremium : 0.03;
```

Comment:
I'm still suspicious of that 3% default. And also I suspect you want pricePremium to be in percent, so it should be divided by 100?

Reply:
pricePremiumFraction is passed on as is to program from here, would I still need to divide by 100? if my intention was 3%...

Reply:
I meant dividing pricePremium, the 0.03 is fine.